File: dotnet/SlimLO/ConversionDiagnostic.cs
using System;

namespace SlimLO;

/// <summary>
/// A single diagnostic entry from a PDF conversion operation.
/// Diagnostics include font substitution warnings, layout issues,
/// and other non-fatal observations collected during conversion.
/// </summary>
public sealed class ConversionDiagnostic
{
    public ConversionDiagnostic(
        DiagnosticSeverity severity,
        DiagnosticCategory category,
        string message,
        string? fontName = null,
        string? substitutedWith = null)
    {
        Severity = severity;
        Category = category;
        Message = message;
        FontName = fontName;
        SubstitutedWith = substitutedWith;
    }

    /// <summary>Severity level (Info or Warning).</summary>
    public DiagnosticSeverity Severity { get; }

    /// <summary>Category (Font, Layout, General).</summary>
    public DiagnosticCategory Category { get; }

    /// <summary>Human-readable diagnostic message.</summary>
    public string Message { get; }

    /// <summary>Font name if this is a font-related diagnostic. Null otherwise.</summary>
    public string? FontName { get; }

    /// <summary>
    /// The font that was used as a substitute, if this is a font substitution warning.
    /// Null if no substitution occurred or if this is not a font diagnostic.
    /// </summary>
    public string? SubstitutedWith { get; }

    public override string ToString() =>
        $"[{Severity}:{Category}] {Message}";
}

File: java/slimlo/src/main/java/com/slimlo/ConversionDiagnostic.java
package com.slimlo;

/**
 * A single diagnostic entry from a PDF conversion operation.
 * Diagnostics include font substitution warnings, layout issues,
 * and other non-fatal observations collected during conversion.
 */
public final class ConversionDiagnostic {

    private final DiagnosticSeverity severity;
    private final DiagnosticCategory category;
    private final String message;
    private final String fontName;
    private final String substitutedWith;

    public ConversionDiagnostic(
            DiagnosticSeverity severity,
            DiagnosticCategory category,
            String message,
            String fontName,
            String substitutedWith) {
        this.severity = severity;
        this.category = category;
        this.message = message;
        this.fontName = fontName;
        this.substitutedWith = substitutedWith;
    }

    /** Severity level (INFO or WARNING). */
    public DiagnosticSeverity getSeverity() {
        return severity;
    }

    /** Category (FONT, LAYOUT, GENERAL). */
    public DiagnosticCategory getCategory() {
        return category;
    }

    /** Human-readable diagnostic message. */
    public String getMessage() {
        return message;
    }

    /** Font name if this is a font-related diagnostic. Null otherwise. */
    public String getFontName() {
        return fontName;
    }

    /** The font that was used as a substitute. Null if no substitution. */
    public String getSubstitutedWith() {
        return substitutedWith;
    }

    @Override
    public String toString() {
        return "[" + severity + ":" + category + "] " + message;
    }
}
